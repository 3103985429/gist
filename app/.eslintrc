{
  // extend root arrays
  "extends": "../.eslintrc.js",
  "parserOptions": {
    "ecmaVersion": 6,
  },
  "env": {
    "browser": false,
    "node": true
  }
}
