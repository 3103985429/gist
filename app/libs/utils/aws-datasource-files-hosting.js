const winston = require('winston');
const aws = require('aws-sdk');
const fs = require('fs');
const _ = require('lodash');

const bucket = process.env.DO_S3_BUCKET;
const endpoint = new aws.Endpoint(process.env.DO_S3_ENDPOINT);
const s3 = new aws.S3({
    endpoint,
    accessKeyId: process.env.DO_ACCESS_KEY_ID,
    secretAccessKey: process.env.DO_SECRET_ACCESS_KEY,
});
const debugS3 = false;

function _uploadAncillaryFile(filePath, fileName, contentType, teamSubdomin, datasetId, callback) {
    const file = fs.createReadStream(filePath);
    const key = getAncillaryAssetKey(teamSubdomin, datasetId, fileName);
    const params = {
        Bucket: bucket,
        Key: key,
        ContentType: contentType,
        Body: file,
        ACL: 'private',
    };

    debug(`Uploading the ancillary file to S3: ${key}`);

    s3.upload(params, function (err) {
        if (err) {
            return callback(err);
        }

        callback();
    });
}

module.exports.uploadAncillaryFile = _uploadAncillaryFile;

function _copyAncillaryFile(team_subdomain, originId, destinationId, fileName, callback) {
    const key = getAncillaryAssetKey(team_subdomain, destinationId, fileName);
    const params = {
        Bucket: bucket,
        CopySource: getAncillaryAssetKey(team_subdomain, originId, fileName, bucket),
        Key: key,
    };

    debug(`Copying the ancillary file from S3: ${key}`);

    s3.copyObject(params, function (err) {
        if (err) {
            winston.error(`S3: error copying datasource: ${err.message}`);
            return callback(err);
        }

        callback();
    });
}

module.exports.copyAncillaryFile = _copyAncillaryFile;

function _uploadDataSource(filePath, newFilename, contentType, teamSubdomin, datasetId, callback) {
    const file = fs.createReadStream(filePath);
    const key = getDatasourceKey(teamSubdomin, datasetId, newFilename);

    const params = {
        Bucket: bucket,
        Key: key,
        ContentType: contentType,
        Body: file,
        ACL: 'private',
    };

    debug(`Uploading the datasource to S3: ${key}`);

    s3.upload(params, function (err) {
        if (err) {
            return callback(err);
        }

        callback();
    });
}

module.exports.uploadDataSource = _uploadDataSource;

function _copySampleDatasource(datasetToDuplicateId, fileName, datasetId, teamSubdomain, callback) {
    const key = getDatasourceKey(teamSubdomain, datasetId, fileName);
    const params = {
        Bucket: bucket,
        CopySource: getDatasourceKey('sampleteam', datasetToDuplicateId, fileName, bucket),
        Key: key,
    };

    debug(`Copying the sample datasource from S3: ${key}`);

    s3.copyObject(params, function (err) {
        if (err) {
            winston.error(`S3: error copying sample datasource: ${err.message}`);
            return callback(err);
        }

        callback();
    });
}

module.exports.copySampleDatasource = _copySampleDatasource;

function _copyDatasource(team_subdomain, fileName, copyFromId, copyToId, callback) {
    const keyFrom = getDatasourceKey(team_subdomain, copyFromId, fileName, bucket);
    const keyTo = getDatasourceKey(team_subdomain, copyToId, fileName);
    const params = {
        Bucket: bucket,
        CopySource: keyFrom,
        Key: keyTo,
    };

    debug(`Copying the datasource from S3: from ${keyFrom} to ${keyTo}`);

    s3.copyObject(params, function (err, data) {
        if (err) {
            winston.error(`S3: error copying datasource: ${err.message}`);
            return callback(err);
        }

        callback();
    });
}

module.exports.copyDatasource = _copyDatasource;

function _getDatasource(description) {
    const key = getDatasourceKey(description._team.subdomain, description.schemaId || description._id, description.fileName);

    const param = {
        Bucket: bucket,
        Key: key,
    };
    debug(`Reading the datasource from S3: ${key}`);

    return s3.getObject(param);
}

module.exports.getDatasource = _getDatasource;

function _getAncillaryFile(description) {
    const key = getAncillaryAssetKey(description._team.subdomain, description._id, description.ancillaryFile);

    const param = {
        Bucket: bucket,
        Key: key,
    };
    debug(`Reading ancillary file from S3: ${key}`);

    return s3.getObject(param);
}

module.exports.getAncillaryFile = _getAncillaryFile;

function keyExists(description, callback = _.noop) {
    return new Promise((resolve, reject) => {
        const key = getDatasourceKey(description._team.subdomain, description.schema_id || description._id, description.fileName);

        const params = {
            Bucket: bucket,
            Key: key,
        };
        debug(`Verifying the datasource from S3: ${key}`);

        s3.headObject(params, (err) => {
            if (err) {
                reject(err);
                return callback(err, false);
            }

            resolve();
            callback(null, true);
        });
    });
}

module.exports.keyExists = keyExists;

function _deleteDataset(description, callback) {
    const team_subdomain = description._team.subdomain;
    const key = getDatasetKey(team_subdomain, description.id);
    const param = {
        Bucket: bucket,
        Prefix: key,
    };

    debug(`Removing the datasource from S3: ${key}`);

    s3.listObjects(param, function (err, data) {
        if (err) {
            return callback(err);
        }

        if (!data.Contents) {
            return callback();
        }

        if (data.Contents.length === 0) {
            return callback();
        }

        const param = { Bucket: bucket };
        param.Delete = { Objects: [] };

        data.Contents.forEach(function (content) {
            param.Delete.Objects.push({ Key: content.Key });
        });

        s3.deleteObjects(param, function (err, data) {
            if (err) {
                return callback(err);
            }

            if (!data.Deleted) {
                return callback();
            }

            if (data.Deleted.length === 1000) {
                return _deleteDataset(description, callback);
            }

            callback();
        });
    });
}

module.exports.deleteDataset = _deleteDataset;

function _deleteTeam(subdomain, callback) {
    const param = {
        Bucket: bucket,
        Prefix: getTeamKey(subdomain),
    };

    s3.listObjects(param, function (err, data) {
        if (err) {
            return callback(err);
        }

        if (!data.Contents) {
            return callback();
        }

        if (data.Contents.length === 0) {
            return callback();
        }

        const param = { Bucket: bucket };
        param.Delete = { Objects: [] };

        data.Contents.forEach(function (content) {
            param.Delete.Objects.push({ Key: content.Key });
        });

        s3.deleteObjects(param, function (err, data) {
            if (err) {
                return callback(err);
            }

            if (!data.Contents) {
                return callback();
            }

            if (data.Contents.length === 1000) {
                return _deleteTeam(subdomain, callback);
            }

            callback();
        });
    });
}

module.exports.deleteTeam = _deleteTeam;

function _deleteObject(key, callback) {
    const params = {
        Bucket: bucket,
        Key: key,
    };

    s3.deleteObject(params, function (err, data) {
        if (err) {
            return callback(err);
        }

        callback(null, data);
    });
}

module.exports.deleteObject = _deleteObject;

function getDatasourceKey(subdomain, datasetId, fileName, bucket = null) {
    return getDatasetKey(subdomain, datasetId, `datasources/${fileName}`, bucket);
}

function getAncillaryAssetKey(subdomain, datasetId, fileName, bucket = null) {
    return getDatasetKey(subdomain, datasetId, `assets/ancillary/${fileName}`, bucket);
}

function getDatasetKey(subdomain, datasetId, resourceName = '', bucket = null) {
    return getTeamKey(subdomain, `datasets/${datasetId}/${resourceName}`, bucket);
}

function getTeamKey(subdomain, resourceName = '', bucket = null) {
    const prefix = bucket ? `${bucket}/` : '';

    return `${prefix}${subdomain}/${resourceName}`;
}

function debug(msg) {
    if (debugS3) {
        winston.debug(msg);
    }
}
