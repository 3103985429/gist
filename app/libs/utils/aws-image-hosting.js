const request = require('request');
var winston = require('winston');
var aws = require('aws-sdk');
var sharp = require('sharp');
var async = require('async');
const _ = require('lodash');

const { VIEW_TYPE_OBJECT_DETAILS, VIEW_TYPE_GALLERY, VIEW_TYPE_TIMELINE } = require('../../config/views.config');

const BUCKET = process.env.DO_S3_BUCKET;
const endpoint = new aws.Endpoint(process.env.DO_S3_ENDPOINT);
const S3 = new aws.S3({
    endpoint,
    accessKeyId: process.env.DO_ACCESS_KEY_ID,
    secretAccessKey: process.env.DO_SECRET_ACCESS_KEY,
    signatureVersion: 'v4',
    region: 'nyc3',
});

module.exports = {
    uploadToS3,
    signedUrlForPutObject,
    getAllIconsForTeam,
    hostImageLocatedAtRemoteURL,
    copyPageAssets,
    copyDatasetAsset,
    getAssetByKey,
};

function uploadToS3(key, response, callback) {

    var payload = {
        Bucket: BUCKET,
        Key: key,
        Body: response,
        ACL: 'public-read',
    };

    S3.upload(payload, function (err, data) {
        if (err) {
            winston.error('AWS S3 write stream error');
            return callback(err);
        }
        return callback(null, data);
    });
}

function _appendKeyToBucket(key) {
    return 'https://' + BUCKET + '.' + process.env.DO_S3_ENDPOINT + '/' + key;
}

function getImageFormat(headers) {
    const imageFormat = _.get(headers, 'content-type', '').split(';')[0].split('/')[1];

    return imageFormat === 'jpg' ? 'jpeg' : imageFormat;
}

function getImageAtUrl(remoteImageSourceURL, callback) {
    var options = {
        url: remoteImageSourceURL,
        encoding: null,
        timeout: 10000,
    };

    request.get(options, function (err, response, data) {
        if ((err && ['ENOTFOUND', 'ETIMEDOUT'].includes(err.code)) || !response) {
            winston.info(`returning url as null, since Could not read the remote image ${remoteImageSourceURL}: ${err}`);
            return callback(null, null);
        }

        if (err) {
            return callback(err, null);
        }

        const imageFormat = getImageFormat(response.headers);

        if (_.get(sharp, `format.${imageFormat}`)) {
            return callback(null, data);
        }

        return callback(null, null);
    });
}

function signedUrlForPutObject(key, fileType, callback) {
    if (fileType.indexOf('svg') >= 0) { //fix space when uploading svg error
        fileType = 'image/svg+xml';
    }

    var params = {
        Bucket: BUCKET,
        Key: key,
        ACL: 'public-read',
        ContentType: fileType,
    };

    S3.getSignedUrl('putObject', params, function (err, signedUrl) {
        callback(err, { putSignedUrl: signedUrl, publicUrl: _appendKeyToBucket(key) });
    });
}

function getAllIconsForTeam(teamSubdomain, callback) {
    S3.listObjects({
        Bucket: BUCKET,
        Prefix: teamSubdomain + '/assets/icon/',
        Marker: teamSubdomain + '/assets/icon/',
    }, function (err, data) {
        if (err) {
            callback(err);
        } else {
            var listOfUrls = [];
            var objects = data.Contents;

            for (var i = 0; i < objects.length; i++) {
                var objKey = objects[i].Key;

                var url = _appendKeyToBucket(objKey);

                listOfUrls.push(url);
            }
            callback(null, listOfUrls);
        }
    });

}

async function _proceedToStreamToHost(imageUrl, resize, folder, docPKey, callback) {
    try {
        const image = sharp(imageUrl);
        const metadata = await image.metadata();

        if (!metadata) {
            winston.info('returning url as null, since Could not read the remote image metadata');
            return callback();
        }

        const { width, hasAlpha } = metadata;
        const size = width < resize.size ? width : resize.size;
        const whiteColorRGB = { r: 255, g: 255, b: 255 };
        const background = hasAlpha ? { ...whiteColorRGB, alpha: 1 } : whiteColorRGB;
        const resizeConfig = { fit: 'contain', background };
        const resizedImage = resize.cropped ? image.resize(size, size, resizeConfig) : image.resize(size);
        const bufferedImage = await resizedImage.toBuffer();
        const keyName = `${folder}${resize.view}/${docPKey}`;

        uploadToS3(keyName, bufferedImage, callback);
    } catch (error) {
        winston.error('has err: ' + error);
        callback(error);
    }
}

function hostImageLocatedAtRemoteURL(folder, remoteImageSourceURL, overwrite, destinationFilenameSansExt, callback) {
    const recommendedSizes = [
        { size: 763, cropped: false, view: VIEW_TYPE_OBJECT_DETAILS },
        { size: 280, cropped: false, view: VIEW_TYPE_GALLERY },
        { size: 155, cropped: true, view: VIEW_TYPE_TIMELINE },
    ];

    getImageAtUrl(remoteImageSourceURL, function (err, imageBuffer) {
        if (err || !imageBuffer) {
            return callback(err);
        } else {
            async.each(recommendedSizes, function (size, cb) {
                if (overwrite) {
                    _proceedToStreamToHost(imageBuffer, size, folder, destinationFilenameSansExt, cb);
                } else {
                    const keyName = `${folder}${size.view}/${destinationFilenameSansExt}`;
                    const params = {
                        Bucket: BUCKET,
                        Key: keyName,
                    };
                    try {
                        S3.headObject(params, function (err, data) {
                            if (err || !data) {
                                winston.error('Head object err: ' + err);
                            } else {
                                winston.info(' File already uploaded and overwrite is false for object with pKey ' + destinationFilenameSansExt);
                                cb();
                            }
                        });
                    } catch (err) {

                        if (err.code === 'ENOTFOUND' || err.code === 'ETIMEDOUT') {
                            _proceedToStreamToHost(imageBuffer, size, folder, destinationFilenameSansExt, cb);
                        } else {
                            winston.error(err);
                            cb(err);
                        }
                    }
                }
            }, callback);
        }
    });

}

function copyPageAssets(args, callback) {
    var params = {
        Bucket: BUCKET,
        CopySource: BUCKET + '/' + args.subdomain + '/websites/' + args.originalWebsiteId + '/pages/' + args.originalPageId + '/' + args.imageType + '/' + args.fileName,
        Key: args.subdomain + '/websites/' + args.websiteId + '/pages/' + args.pageId + '/' + args.imageType + '/' + args.fileName,
        ACL: 'public-read',
    };

    S3.copyObject(params, function (err) {
        if (err) {
            winston.error('\' error copying page asset with params: \'', params);
            return callback(err);
        } else {
            callback();
        }
    });
}

async function copyDatasetAsset({ fromSubdomain, toSubdomain, fromId, toId, source, fileName }, callback) {
    const getUrl = ({ id, BUCKET, subdomain = fromSubdomain }) =>
        `${BUCKET ? `${BUCKET}/` : ''}${subdomain}/datasets/${id}/${source}/${fileName}`;
    const copySource = getUrl({ id: fromId, BUCKET });
    const key = getUrl({ id: toId, subdomain: toSubdomain });

    const params = {
        Bucket: BUCKET,
        CopySource: copySource,
        Key: key,
        ACL: 'public-read',
    };

    try {
        await S3.copyObject(params).promise();

        callback(null);
    } catch (e) {
        winston.error(`S3: error copying dataset asset: ${e}`);

        callback(e);
    }
}

function getAssetByKey(key, callback) {
    const params = {
        Bucket: BUCKET,
        Key: key,
    };

    S3.getObject(params, function (err, data) {
        if (err) {
            return callback(err);
        } else {
            return callback(null, data);
        }
    });
}
