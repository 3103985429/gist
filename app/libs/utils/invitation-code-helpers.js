const crypto = require('crypto');

/**
 * Returns a sha512 hash of the given string.
 * @param confirmation_code - Code to hash
 * @return
 */
module.exports.createHashWithString = confirmation_code => {
    // Create a new hash, and then update it and return the digest
    const hash = crypto.createHash('sha512');
    hash.update(confirmation_code);
    return hash.digest('hex');
};

/**
 * Generates a random short code.
 * @param codeSize - Size in bytes to generate a code
 * @return
 */
module.exports.generateRandomCode = codeSize => {
    // Get a random buffer
    const randomBuffer = crypto.randomBytes(codeSize);
    // Return the random buffer to string
    return randomBuffer.toString('hex');
};
