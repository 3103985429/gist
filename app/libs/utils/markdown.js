var fs = require('fs');
var marked = require('marked');

module.exports.getMarkdown = function (filePath) {
    var rawMarkdown = fs.readFileSync(filePath, 'utf-8');
    return rawMarkdown;
};

module.exports.parseMarkdown = function (markdown) {
    var options = {
        renderer: new marked.Renderer(),
        gfm: true,
        tables: true,
        breaks: false,
        pedantic: false,
        sanitize: true,
        smartLists: true,
        smartypants: false
    };
    marked.setOptions(options);

    var lexer = new marked.Lexer(options);
    var parsedBody = marked(markdown);
    return parsedBody;
};

module.exports.replaceCharts = function (parsed) {
    var chartRE = /((?:<p>)?&lt;chart data-source-key=&quot;(?=\S*)([a-zA-Z\d_-]*)&quot; data-chart-type=&quot;(?=\S*)([a-zA-Z\d-]*)&quot; class=&quot;chart&quot; data-revision=&quot;(\d+)&quot; \/&gt;(?:<\/p>)?)/g;
    var parsedReplaced = parsed.replace(chartRE, '<div id="$2" data-source-key="$2" data-chart-type="$3" class="$3 chart" data-revision="$4"></div>');
    return parsedReplaced;
};
