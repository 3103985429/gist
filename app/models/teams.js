const integerValidator = require('mongoose-integer');
const winston = require('winston');
const { isEmpty } = require('lodash');

const mongoose_client = require('./mongoose_client');

const mongoose = mongoose_client.mongoose;
const Schema = mongoose.Schema;

const team_scheme = Schema({
    title: String,
    subdomain: { type: String, unique: true },
    description: String,
    font: { type: String, default: 'centrano2' },
    logo: String,
    logo_header: String,
    externalWebsite: String,
    colorPalette: {
        type: Array,
        default: [
            '#1B6DFC',
            '#0CB7FA',
            '#FA79FC',
            '#FB3705',
            '#F9D307',
            '#FA9007',
            '#7D5807',
            '#B2E606',
            '#069908',
            '#7B07FD',
        ],
    },
    superTeam: Boolean,
    admin: [{ type: Schema.Types.ObjectId, ref: 'User' }],
    datasourceDescriptions: [{ type: Schema.Types.ObjectId, ref: 'DatasourceDescription' }],
    pages: [{ type: Schema.Types.ObjectId, ref: 'Page' }],
    sites: [{ type: Schema.Types.ObjectId, ref: 'Website' }],

    isEnterprise: Boolean, //if true, means it has custom view - TODO consolidate with hasEnterpriseLicense
    customViewModal: Boolean, //if true, custom view has custom view settings modal - (rhodiumgroup)
    hasEnterpriseLicense: { type: Boolean, default: false },

    google_analytics: String,
    header_script: String,
    footer_script: String,
    brandColor: {
        label: { type: String, default: '#2A2A2A' },
        select: { type: String, default: '#2A2A2A' },
    },

}, { timestamps: true });

const modelName = 'Team';

team_scheme.plugin(integerValidator);

const deepPopulate = require('mongoose-deep-populate')(mongoose);

team_scheme.plugin(deepPopulate, { whitelist: ['datasourceDescriptions.author', 'datasourceDescriptions.updatedBy', 'datasourceDescriptions.createdAt'] });

team_scheme.pre('save', function (next) {
    this._wasNew = this.isNew;
    next();
});

team_scheme.methods.notifyNewTeamCreation = function () {
    if (this._wasNew) {
        this.populate('admin', function (err, docPopulatedWithAdmin) {
            if (err || isEmpty(docPopulatedWithAdmin.admin)) {
                winston.error('Team created with error err:' + err);
            }
        });
    }
};

team_scheme.statics.getTeamEntries = async (subdomain, user, vizName) => {
    const baseQuery = { subdomain };

    if (process.env.NODE_ENV === 'enterprise' && !process.env.SUBTEAMS) {
        baseQuery.subdomain = process.env.SUBDOMAIN;
    }

    const populateQuery = [
        {
            path: 'datasourceDescriptions',
            match: require('./descriptions').buildTeamPageQuery(user, vizName),
            select: 'description uid urls title importRevision updatedAt createdAt previous_datasets updatedBy author brandColor fe_views fe_filters.default banner connection',
            populate: [
                {
                    path: 'updatedBy',
                    select: 'firstName lastName',
                },
                {
                    path: 'author',
                    select: 'firstName lastName',
                },
                {
                    path: 'previous_datasets',
                    select: 'createdAt -_id',
                },
            ],
            options: {
                sort: {
                    createdAt: -1,
                },
            },
        },
        {
            path: 'pages',
            match: require('./pages').buildTeamPageQuery(user),
            select: 'pageTitle slug parsedSummary url thumbnail image updatedBy createdBy createdAt updatedAt createdAt',
            populate: [
                {
                    path: 'createdBy',
                    select: 'firstName lastName -_id',
                },
                {
                    path: 'website',
                    select: 'slug -_id',
                },
            ],
            options: {
                sort: {
                    createdAt: -1,
                },
            },
        },
    ];

    return team.find(baseQuery).populate(populateQuery);
};

const team = mongoose.model(modelName, team_scheme);

team.GetTeams = function (fn) {
    mongoose_client.WhenMongoDBConnected(function () {
        team.find({}, function (err, teams) {
            if (err) {
                fn(err);
            } else {
                fn(null, teams);
            }
        });
    });
};

module.exports = team;
