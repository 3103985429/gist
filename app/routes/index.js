const { defaultTo } = require('lodash');
const helmet = require('helmet');
const express = require('express');

const { errorLogger } = require('../utils/winston');
const { checkCors, API_ALLOWED } = require('./permissions');
const { version } = require('../../package.json');

const PROTOCOL = process.env.USE_SSL === 'true' ? 'https' : 'http';
const HOST = defaultTo(process.env.HOST, 'localhost:9080');
const ROOT_DOMAIN = `${PROTOCOL}://app.${HOST}`;
const API_VERSION = 'v1';

const _mountRoutes_monitoring = function (app) {
    app.get('/_ah/health', function (req, res) {
        res.set('Content-Type', 'text/plain');
        res.status(200).send('ok');
    });
};

const _mountRoutes_errorHandling = function (app) {
    // Add the error logger after all middleware and routes so that
    // it can log errors from the whole application. Any custom error
    // handlers should go after this.
    errorLogger(app);

    // eslint-disable-next-line no-unused-vars
    app.use(function (err, req, res, next) {
        const status = err.status || 404;
        const [title, message] = (status === 404) ?
            ['Oops!', 'The page you are looking for doesn\'t exist or has been moved.'] :
            ['Uh oh.', 'Something went wrong\u2026'];
        const error = process.env.NODE_ENV === 'development' ? err.toString() : `Error: ${status}`;

        res.status(status).render('error', {
            env: process.env,
            title,
            message,
            err: error,
            has_google_analytics: !!process.env.GOOGLE_ANALYTICS_ID,
        });
    });

    app.use(function (req, res) {
        res.status(404);
        res.render('error', {
            env: process.env,
            title: 'Oops!',
            message: 'The page you are looking for doesn\'t exist or has been moved.',
            err: 'Error: 404',
            has_google_analytics: !!process.env.GOOGLE_ANALYTICS_ID,
        });
    });
};

const _mountRoutes_endPoints = function (app) {
    app.all('*', function ({ subdomains, url }, res, next) {
        const redirectToRootDomainPatterns = [
            /^\/dashboard/,
            /^\/login/,
            /^\/signup/,
            /^\/auth/,
            /^\/product/,
            /^\/news/,
            /^\/pricing/,
            /^\/plans/,
        ];
        const isEnterprise = process.env.NODE_ENV === 'enterprise';
        const isRootDomain = subdomains.length === 1 && ['app', 'www'].includes(subdomains[0]);

        if (!isEnterprise && !isRootDomain && redirectToRootDomainPatterns.some(pattern => pattern.test(url))) {
            return res.redirect(ROOT_DOMAIN + url);
        }

        next();
    });

    app.use('/robots.txt', express.static(__dirname + '/../../robots.txt'));

    app.use('/', require('./homepage'));

    app.use('/s', require('./shared_pages'));
    app.use('/presentation', require('./stories'));
    app.use(`/json-api/${API_VERSION}`, checkCors(API_ALLOWED), function (req, res, next) {
        req.apiVersion = API_VERSION;
        next();
    }, require('./json-api/router'));
    app.use('/auth', helmet.frameguard(), checkCors(), require('./auth'));
    app.use('/login', helmet.frameguard(), checkCors(), function (req, res) {
        res.redirect('/auth/login');
    });
    app.use('/signup', helmet.frameguard(), checkCors(), require('./signup'));
    app.use('/reset', helmet.frameguard(), checkCors(), function (req, res) {
        // TODO: Perform an jwt auth here instead of in account.js / or better yet auth in both!
        res.render('auth/password', {
            env: process.env,
        });
    });

    app.use('/dashboard', helmet.frameguard(), checkCors(), require('./dashboard'));

    app.use('/api', helmet.frameguard(), require('./api'));

    app.use('/account', checkCors(), require('./account'));
    app.use('/edit-visualization', require('./external/editVisualization'));
    app.use('/create-visualization', require('./external/createVisualization'));
    app.use('/visualization-list', require('./external/visualizationList'));

    app.use('/', require('./articles'));
    app.use('/', checkCors(API_ALLOWED), require('./views').router);
};

module.exports.MountRoutes = function (app) {
    app.get('/env', function (req, res) {
        const obj = {
            node_env: process.env.NODE_ENV,
            host: HOST,
            s3Bucket: process.env.DO_S3_BUCKET,
            s3Domain: process.env.DO_S3_ENDPOINT,
            subdomain: process.env.SUBDOMAIN,
            authProtocol: process.env.AUTH_PROTOCOL,
            hasSubteams: process.env.SUBTEAMS,
            git: {
                branch: process.env.BRANCH_NAME,
                commit: process.env.COMMIT_HASH,
            },
            version,
            insight: process.env.INSIGHT,
            hide_business_links: process.env.HIDE_BUSINESS_LINKS,
            hasSmartsheetInfo: !!process.env.SMARTSHEET_CLIENT_ID,
            hasPipedriveInfo: !!process.env.PIPEDRIVE_CLIENT_ID,
            hasDatadotworldInfo: !!process.env.DATA_DOT_WORLD_CLIENT_ID,
            hasSalesforceInfo: !!process.env.SALESFORCE_CONSUMER_KEY
        };

        return res.json(obj);
    });

    _mountRoutes_monitoring(app);
    _mountRoutes_endPoints(app);
    _mountRoutes_errorHandling(app);
};
