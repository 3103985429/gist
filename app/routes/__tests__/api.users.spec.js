import request from 'supertest';

import app from '../../../app';
import User from '../../../app/models/users';
import Tokens from '../../../app/models/tokens';
import { hydrateDb, insert, getId } from '../../../internals/testing/backend/utils';
import { team2 } from '../../../internals/testing/backend/fixtures/teams';
import jwt from 'jsonwebtoken';
import { apiKey1 } from '../../../internals/testing/backend/fixtures/apiKeys';

describe('Users API', () => {
    let appRequest;

    beforeEach(async () => {
        await hydrateDb();
        appRequest = request(app);
    });

    describe('Creating a new account', () => {
        it('should create a new user', async () => {
            const token = await insert(Tokens, { token: jwt.sign({ userId: apiKey1.user }, process.env.SESSION_SECRET, { expiresIn: '24h' }) });

            const reqBody = {
                email: 'user+newuser@schemadesign.com',
                provider: 'local',
                tokenId: token.id,
                _team: { title: 'wizards', subdomain: 'wizards' },
                firstName: 'Mary',
                lastName: 'Totter',
                password: 'Dogwarts123',
                verifyPassword: 'Dogwarts123',
            };

            const response = await appRequest.post('/api/user').send(reqBody);

            expect(response.statusCode).toBe(200);
            delete response.body.hash;
            delete response.body.salt;
            delete response.body.inviteToken;
            delete response.body.defaultLoginTeam;
            delete response.body._team;
            expect(response).toMatchSnapshot();
        });

        it('should get a user', async () => {
            const userData = { email: 'user+200@schemadesign.com', provider: 'local' };
            const newUser = await insert(User, userData);
            const response = await appRequest.get(`/api/user/${getId(newUser)}`);
            expect(response.statusCode).toBe(200);
            expect(response).toMatchSnapshot();
        });

        it('should update a user (invited to a team)', async () => {
            const userData = { email: 'user+newuser@schemadesign.com', provider: 'local' };
            const newUser = await insert(User, userData);
            const reqBody = {
                ...newUser._doc,
                _team: team2,
                firstName: 'Oryx',
                lastName: 'Crake',
                password: 'Password123',
                verifyPassword: 'Password123',
            };
            const response = await appRequest.put(`/api/user/${getId(newUser)}`).send(reqBody);
            delete response.body.hash;
            delete response.body.salt;
            expect(response.statusCode).toBe(200);
            delete response.body.hash;
            delete response.body.salt;
            expect(response).toMatchSnapshot();
        });
    });
});
