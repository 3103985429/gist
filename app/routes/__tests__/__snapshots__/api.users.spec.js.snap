// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Users API Creating a new account should create a new user 1`] = `
Object {
  "body": Object {
    "_articleEditors": Array [],
    "_articleViewers": Array [],
    "_editors": Array [],
    "_siteEditors": Array [],
    "_siteViewers": Array [],
    "_viewers": Array [],
    "activated": true,
    "active": true,
    "canCreateNewArticle": Array [],
    "canCreateNewSite": Array [],
    "canCreateNewViz": Array [],
    "email": "user+newuser@schemadesign.com",
    "firstName": "Mary",
    "inviter": "5c6acadf5fbcd004266b2f8c",
    "lastName": "Totter",
    "lockUntil": 1,
    "loginAttempts": 0,
    "provider": "local",
    "role": "editor",
    "sampleImported": false,
    "signedUp": true,
  },
  "statusCode": 200,
}
`;

exports[`Users API Creating a new account should get a user 1`] = `
Object {
  "body": Object {
    "_articleEditors": Array [],
    "_articleViewers": Array [],
    "_editors": Array [],
    "_siteEditors": Array [],
    "_siteViewers": Array [],
    "_team": Array [],
    "_viewers": Array [],
    "activated": false,
    "active": true,
    "canCreateNewArticle": Array [],
    "canCreateNewSite": Array [],
    "canCreateNewViz": Array [],
    "email": "user+200@schemadesign.com",
    "lockUntil": 1,
    "loginAttempts": 0,
    "provider": "local",
    "sampleImported": false,
    "signedUp": false,
    "team": Array [],
  },
  "statusCode": 200,
}
`;

exports[`Users API Creating a new account should update a user (invited to a team) 1`] = `
Object {
  "body": Object {
    "_articleEditors": Array [],
    "_articleViewers": Array [],
    "_editors": Array [],
    "_siteEditors": Array [],
    "_siteViewers": Array [],
    "_team": Array [
      "5a42fba8999d2a26d50fea64",
    ],
    "_viewers": Array [],
    "activated": true,
    "active": true,
    "canCreateNewArticle": Array [],
    "canCreateNewSite": Array [],
    "canCreateNewViz": Array [],
    "defaultLoginTeam": "5a42fba8999d2a26d50fea64",
    "email": "user+newuser@schemadesign.com",
    "firstName": "Oryx",
    "lastName": "Crake",
    "lockUntil": 1,
    "loginAttempts": 0,
    "provider": "local",
    "sampleImported": false,
    "signedUp": true,
  },
  "statusCode": 200,
}
`;
